### DISCUSSION TITLE
Webview2 New Popup window requested but opened in external browser window and not focused in UWP app

### DISCUSSION THREAD

**Discussion by Suresh-BabuLV** [MicrosoftEdge/WebView2Feedback#3954] [Q&A]
Hi Team,

We have developed one UWP app that app need to learn courses when we click on learn course window that window open in new popup window
**We are not able to see that new course window in UWP but try to see that new window to click Task manager to open that window in windows OS, It is not possible to see that New popup (Course) window in Kiosk Systems **

I have tried below code to open new course window on same(Main) window but blank window is opened

 private void CoreWebView2_NewWindowRequested(Microsoft.Web.WebView2.Core.CoreWebView2 sender, Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventArgs args)
        {
            args.Handled = true;
            args.NewWindow = (Microsoft.Web.WebView2.Core.CoreWebView2)sender;
            args.GetDeferral().Complete();
        }
Requirement:1. Open new popup course window in Same window 
                      2.Focus that popup window on same(Main window)

I have tried alternate ways but doesn't work for me, Please provide your valuable suggestions to resolve this issue

Thanks in Advance

Regards,
Suresh.

**Comment by victorthoang** [comment]
Hello Suresh, 

Could you please provide some of the following information while we get back to you:
 
SDK:
Runtime:
Framework:
Sample code to repo and examine the issue?:
Works in Edge without WV2?:

Thanks!

**Comment by david-risney** [comment]
The [documentation for NewWindowRequestedEventArgs.NewWindow](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2newwindowrequestedeventargs#put_newwindow) says:

> The NewWindow property should be set to a CoreWebView2 that has not been navigated previously. Don't use methods that cause navigation or interact with the DOM on this CoreWebView2. Setting event handlers, changing Settings properties, or other methods are fine to call. Changes to settings should be made before put_NewWindow is called to ensure that those settings take effect for the newly setup WebView. Once the NewWindow is set the underlying web contents of this CoreWebView2 will be replaced and navigated as appropriate for the new window. After setting new window it cannot be changed and error will be return otherwise.

You are providing the same CoreWebView2 that has made the request and so it has already navigated. 

Instead, can you create a new CoreWebView2? If you intend for it to show in the same window, you can replace the old WebView2 control with a new WebView2 control and use the new WebView2 control's CoreWebView2 for the new window.

Alternatively, you might also try cancelling the NewWindowRequested event (by setting Handled to true, and NewWindow to null), and then manually navigating the sender CoreWebView2 to the URI of the NewWindow request. The NewWindowRequested event has more requirements about the NewWindow because it has to connect the JS engine in the sender to the JS engine in the new window's. But if you are just discarding the previous CoreWebView2, then you could skip the NewWindow part and manually navigate the original CoreWebView2.

**Comment by Suresh-BabuLV** [comment]
Great Thanks @david-risney,

I have tried to setting handled property is true and args.Newwindow is null is getting error it's not accessible parameter.

When New window request event triggers time to first created webview2 visibility setting as Visibily.collapsed
And create new webview2 and bind the source is args.uri to append to the grid. but New window is opened the behind the main window but not focused on the main window is not working as expected.

Please provide more suggestions to resolve the Issue.


Thanks in Advance.

Regards,
Suresh.

**Comment by david-risney** [comment]
> I have tried to setting handled property is true and args.Newwindow is null is getting error it's not accessible parameter.

Can you please paste the actual error message and the code around the error? I'm not sure what you are saying had the error or what the error was.

>  but New window is opened the behind the main window but not focused on the main window is not working as expected.

Also not sure exactly what you are saying here. Is there a new window, or do you have two WebView2 controls on the same grid? Are you saying that one has focus but it is not the one that is displayed?

**Comment by Suresh-BabuLV** [comment]
Great Thanks @david-risney 

Please find the below Error 
![image](https://user-images.githubusercontent.com/102970058/169240322-7f12b1ac-909e-4d00-b4a4-355162748178.png)

We have only one Webview2.
Requirement: Need to focus(Active) on New popup window on top of that main window

Q. Also not sure exactly what you are saying here. Is there a new window, or do you have two WebView2 controls on the same grid? Are you saying that one has focus but it is not the one that is displayed?

A. We have only one Webview2. When we click on Learn course which is opening new popup window behind the main window

After your suggestion tried that second webview2 created and bind the source into that webview2 that one also not working as expected.

Expected Behavior: New popup window need to open and active(Focus) on Main window

Please Provide More suggestions to resolve this issue.

Thanks in Advance

Regards,
Suresh.

**Comment by Suresh-BabuLV** [comment]
Thanks @david-risney,

I have provided the screenshot of error.
Please provide your valuable inputs to resolve this issue


Regards,
Suresh.

**Comment by david-risney** [comment]
Oh wow there's an AccessViolation when setting the property to null.  That is not expected, but thankfully it defaults to null and we don't need to set the property to null.

I've made a sample app https://github.com/david-risney/WebView2Samples/tree/uwp-new-window-playground (its in the uwp-new-window-playground branch) that shows different ways of handling the NewWindowRequested event. Are you able to reproduce your issue in this sample or does the sample help answer your question? How are you opening your new window?

You said your app is UWP, is it also WinUI3?

**Comment by Suresh-BabuLV** [comment]
@david-risney, Great Thanks.

I sincerely appreciate your help.

I have tried as your suggestions it's working as expected but here getting one more issue.
The issue was event(like button) controls are not working on webview2right.
(Example: New course window opened in the same webview2 by using webview2right
The user controls are not clickable like submit questions and click next session on same window not able to click on that button events.)
Please provide your valuable suggestion.


Thanks in Advance

Regards,
Suresh.

**Comment by david-risney** [comment]
I'm using the sample app I referenced above with https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input in both the left and right WebView2s and I don't see issues with any of the HTML controls. Are you seeing the problem in the sample app and if so which URI are you on?

**Comment by Suresh-BabuLV** [comment]
Great Thanks @david-risney,

I have tested with this URL ([Google website](https://www.google.com/)) 
your sample code was working user controls .Yes right side webview2(New popup window) is working fine.
But when we used to embed the second webview2 user controls are not working on our project.

Please suggest and guide to me. What can i do for enable to user controls are working whole embed second(right side) webview2.


Regards,
Suresh.

**Comment by david-risney** [comment]
I'm not sure. It sounds like you have a problem with either input or focus. @bradp0721 may be able to provide some suggestions.

**Comment by Suresh-BabuLV** [comment]
Hi  @bradp0721, @david-risney.

Please provide your valuable suggestions and guide to inject JavaScript or any input html elements to enable the input controls or focus input controls in webview2.


Thanks in Advance.


Regards,
Suresh.

**Comment by bradp0721** [comment]
It sounds like input is working correctly in the sample app but not in your project? Is there something your project is doing differently? Can you try loading the same (or similar) html in into the sample app and see if that reproduces the input issue?
